cur-lib/cur/stdlib/sugar.rkt

```diff
@@ -210,25 +244,32 @@
                                           #`(#,@#'((x ty_in1) ...)
                                              #,@x+τs))
                                         #'(([xpat xpatτ] ...) ...))
-     #:with (ty-to-match/ ...) (stx-map ; 'recur is for termination check
-                                (λ (t) (syntax-property t 'recur #t))
-                                #'(ty-to-match ...))
+     ;; mark the specified argument as decreasing - if it's the last one we also
+     ;; indicate it so that we can fail properly
+     #:with (ty-to-match/ ...) (for/list ([ty (attribute ty-to-match)]
+                                          [i (in-naturals)])
+                                 (if (= i (stx->datum (attribute decreasing-arg)))
+                                     (syntax-property (syntax-property ty 'decreasing #t) 'last-decreasing-check
+                                                      (= i (sub1 (length (attribute ty-to-match)))))
+                                     ty))
      [([x+pat ≫ x+pat- : x+patτ] ...)
       ([y ≫ y*- : ty_in2] ...
        ;; for now, assume recursive references are fns
        [name ≫ name- : (Π [x : ty_in1] ... [x0 : ty-to-match/] ... [y : ty_in2] ... ty_out)])
       ⊢ [body ≫ body- ⇐ ty_out]
       #:where typecheck-relation
-           (λ (t1 t2)
-             (and (old-typecheck-relation t1 t2)
-                  (or (not (syntax-property t2 'recur))
-                      (and (syntax-property t2 'recur)
-                           (let ([t1-recur-ok? (syntax-property t1 'recur)])
-                             (begin0 t1-recur-ok?
-                               (unless t1-recur-ok?
-                                 (fprintf (current-error-port)
-                                          "Failed termination check for arg of type ~a:\n"
-                                          (stx->datum (resugar-type t1))))))))))
+      (λ (t1 t2)
+        (and (old-typecheck-relation t1 t2)
```

Comment:
Where is `old-type-check-relation` defined?

Reply:
It's implicit magic :/ Open to alternative designs.